client/app/components/EditParameterSettingsDialog.jsx

```diff
@@ -0,0 +1,214 @@
+import { includes, words, capitalize, clone, isNull } from 'lodash';
+import React, { useState, useEffect } from 'react';
+import PropTypes from 'prop-types';
+import Modal from 'antd/lib/modal';
+import Form from 'antd/lib/form';
+import Checkbox from 'antd/lib/checkbox';
+import Select from 'antd/lib/select';
+import Input from 'antd/lib/input';
+import Divider from 'antd/lib/divider';
+import { wrap as wrapDialog, DialogPropType } from '@/components/DialogWrapper';
+import { Query } from '@/services/query';
+
+const { Option } = Select;
+const formItemProps = { labelCol: { span: 6 }, wrapperCol: { span: 16 } };
+
+function useQuerySearch() {
+  const [searchTerm, setSearchTerm] = useState(); // search term
+  const [searchResults, setSearchResults] = useState([]); // search results
+
+  useEffect(() => {
+    if (!searchTerm || searchTerm.length < 3) {
+      return;
+    }
+
+    Query.query({ q: searchTerm }, ({ results }) => {
+      setSearchResults(results);
+    });
+  }, [searchTerm]); // search when term changes
+
+  return [searchResults, setSearchResults, setSearchTerm];
+}
+
+function useParamSave(onConfirm, onClose) {
+  const [saving, setSaving] = useState(false);
+
+  const save = (param) => {
+    setSaving(true);
+    onConfirm(param)
+      .then(() => {
+        onClose(param);
+      }).finally(() => {
+        setSaving(false);
+      });
+  };
+
+  return [saving, save];
+}
+
+function getDefaultTitle(text) {
+  return capitalize(words(text).join(' ')); // humanize
+}
+
+function NameInput({ name, onChange, existingParams, setValidation }) {
+  const alreadyExists = includes(existingParams, name);
+  let helpText = 'Choose a keyword for this parameter';
+
+  if (alreadyExists) {
+    helpText = 'Parameter with this name already exists';
+    setValidation(false);
+  } else if (!name) {
+    helpText = 'Choose a keyword for this parameter';
+    setValidation(false);
+  } else {
+    helpText = `This is what will be added to your query editor {{ ${name} }}`;
+    setValidation(true);
+  }
+
+  return (
+    <Form.Item
+      required
+      label="Keyword"
+      help={helpText}
+      validateStatus={alreadyExists ? 'error' : ''}
+      {...formItemProps}
+    >
+      <Input onChange={e => onChange(e.target.value)} />
+    </Form.Item>
+  );
+}
+
+NameInput.propTypes = {
+  name: PropTypes.string.isRequired,
+  onChange: PropTypes.func.isRequired,
+  existingParams: PropTypes.arrayOf(PropTypes.string).isRequired,
+  setValidation: PropTypes.func.isRequired,
+};
+
+function EditParameterSettingsDialog(props) {
+  const [param, setParam] = useState(clone(props.parameter));
+  const [querySearchResults, setQuerySearchResults, setQuerySearchTerm] = useQuerySearch();
+  const [saving, save] = useParamSave(props.onConfirm, props.dialog.close);
+  const [isNameValid, setIsNameValid] = useState(true);
+  const isNew = !props.parameter.name;
+
+  // get query name by id
+  useEffect(() => {
```

Comment:
I think this should be above `isNew`. I recall reading that all hooks should be defined first in the function.

Reply:
(there is an ESLint plugin for this)